haystack/components/agents/agent.py

```diff
@@ -361,6 +365,46 @@ def _initialize_fresh_execution(
             tool_invoker_inputs=tool_invoker_inputs,
         )
 
+    def _validate_tools_against_state_schema(self, tools: ToolsType) -> None:
+        """
+        Validate that tools' state mappings reference valid state schema keys.
+
+        :param tools: Tools to validate against the state schema.
+        :raises ValueError: If any tool references invalid state keys.
+        """
+        for tool in flatten_tools_or_toolsets(tools):
+            # Validate inputs_from_state keys exist in state schema
+            if tool.inputs_from_state:
+                for state_key in tool.inputs_from_state.keys():
+                    if state_key not in self.state_schema:
+                        available_keys = set(self.state_schema.keys())
+                        tool_param = tool.inputs_from_state[state_key]
+                        raise ValueError(
+                            f"Tool '{tool.name}' reads from unknown state key '{state_key}'.\n"
+                            f"Available state keys: {available_keys}\n\n"
+                            f"To fix this:\n"
+                            f"1. Add '{state_key}' to your Agent's state_schema:\n"
+                            f"   Agent(..., state_schema={{'{state_key}': {{'type': YourType}}, ...}})\n"
+                            f"2. Or map from an existing state key to tool parameter '{tool_param}':\n"
+                            f"   Tool(..., inputs_from_state={{'existing_state_key': '{tool_param}'}})"
+                        )
+
+            # Validate outputs_to_state keys exist in state schema (only when no handler)
+            if tool.outputs_to_state:
+                for state_key, config in tool.outputs_to_state.items():
+                    if config.get("handler") is None and state_key not in self.state_schema:
+                        available_keys = set(self.state_schema.keys())
+                        raise ValueError(
+                            f"Tool '{tool.name}' writes to unknown state key '{state_key}'.\n"
+                            f"Available state keys: {available_keys}\n\n"
+                            f"To fix this:\n"
+                            f"1. Add '{state_key}' to your Agent's state_schema:\n"
+                            f"   Agent(..., state_schema={{'{state_key}': {{'type': YourType}}, ...}})\n"
+                            f"2. Or provide a handler to create the key dynamically:\n"
+                            f"   Tool(..., outputs_to_state={{'{state_key}':"
+                            f" {{'source': 'output', 'handler': your_handler}}}})"
```

Comment:
This isn't correct. The handler does not create the state key dynamically. It handles how the merge operation should occur for an existing state key. For example, if we are writing to the state key `documents: list[Document]`, the handler determines how the new value being written is merged with the existing value in state. 

So the handler doesn't affect the key mapping at all.